src/plugins/validate/index.js

```diff
@@ -0,0 +1,99 @@
+import * as selectors from "./selectors"
+import * as actions from "./actions"
+import traverse from "traverse"
+import {createSelector} from "reselect"
+import { fromJS } from "immutable"
+import debounce from "lodash/debounce"
+
+export default function SemanticValidatorsPlugin({getSystem}) {
+
+  const debAll = debounce((system) => system.validateActions.all(), 300)
+  const traverseOnce = makeTraverseOnce(getSystem)
+
+  return {
+    fn: {
+      traverse,
+      traverseOnce,
+    },
+    statePlugins: {
+      spec: {
+        selectors: {
+          jsonAsJS: createSelector(
+            state => state.get("resolvedSpec"),
+            state => state.get("json"),
+            (a,b) => (a || b).toJS()
+          ),
+          definitions: createSelector(
+            state => state.getIn(["json", "definitions"]),
+            (defs) => defs || fromJS({})
+          ),
+        },
+        wrapActions: {
+          validateSpec: (ori, system) => (...args) => {
+            ori(...args)
+            debAll(system)
+          }
+        }
+      },
+      validate: {
+        selectors,
+        actions,
+      },
+    }
+  }
+}
+
+function makeTraverseOnce(getSystem) {
+  let traversers = {}
+  let results = {}
+  let deferred = null
+  let timer = null
+
+  const debTraverse = debounce(() => {
+    timer = new (getSystem().fn.Timer)("traverseOnce")
+
+    // Setup collections
+    for(let name in traversers) {
+      results[name] = []
+    }
+
+    const json = getSystem().specSelectors.jsonAsJS()
+    timer.mark("got json")
+
+    getSystem().fn.traverse(json)
+      .forEach(function() { // Remember: this cannot be a fat-arrow function, because we need to read "this"
+        for(let name in traversers) {
+          const fn = traversers[name]
+          const fnRes = fn(this)
```

Comment:
We should use a filter instead of relying on `fn` to return the node.
We should not pass in the traverse instance directly instead we should use a simple interface for future support...
eg:
```js
const shouldAdd = fn({ path: this.path, value: this.value }) // Note  we should probably support some sort of "traversal" here, ie: how does one navigate to the parent? this.parent exists, but it would yield another traverse obj... blarg 
if(shouldAdd) { ... }
```